keps/sig-autoscaling/20181106-in-place-update-of-pod-resources.md

```diff
@@ -0,0 +1,350 @@
+
+The **core idea** behind the proposal is to make PodSpec mutable with regards to
+Resources, denoting **desired** resources.
+Additionally, PodStatus is extended to provide information about **actual**
+resource allocation.
+
+This document builds upon [proposal for live and in-place vertical scaling][] and
+[Vertical Resources Scaling in Kubernetes][].
+
+[proposal for live and in-place vertical scaling]: https://github.com/kubernetes/community/pull/1719
+[Vertical Resources Scaling in Kubernetes]: https://docs.google.com/document/d/18K-bl1EVsmJ04xeRq9o_vfY2GDgek6B6wmLjXw-kos4/edit?ts=5b96bf40
+
+## Motivation
+
+Resources allocated to a Pod's Container can require a change for various reasons:
+* load handled by the Pod has increased significantly and current resources are
+  not sufficient,
+* load has decreased significantly and allocated resources are unused and wasted,
+* resources have simply been set improperly.
+
+Currently, changing resource allocation requires the Pod to be recreated since
+the PodSpec's Container Resources is immutable.
+
+While many stateless workloads are designed to withstand such a disruption, some
+are more sensitive, especially when using low number of Pod replicas.
+
+Moreover, for stateful or batch workloads, a Pod restart is a serious
+disruption, resulting in lower availability or higher cost of running.
+
+Allowing Resources to be changed without recreating the Pod or restarting the
+Containers addresses this issue directly.
+
+### Goals
+
+* Primary: allow to change Pod resource requests & limits without restarting its
+  Containers.
+* Secondary: allow actors (users, VPA, StatefulSet, JobController) to decide
+  how to proceed if in-place resource resize is not possible.
+* Secondary: allow users to specify which Pods and Containers can be resized
+  without a restart.
+
+### Non-Goals
+
+The explicit non-goal of this KEP is to avoid controlling full life-cycle of a
+Pod which failed in-place resource resizing. This should be handled by actors
+which initiated the resizing.
+
+Other identified non-goals are:
+* allow to change Pod QoS class without a restart,
+* to change resources of Init Containers without a restart,
+* updating extended resources or any other resource types besides CPU, memory.
+
+## Proposal
+
+### API Changes
+
+PodSpec becomes mutable with regards to Container resources requests and limits.
+Additionally, PodSpec becomes a Pod subresource to allow fine-grained access control.
+
+PodStatus is extended with information about actually allocated Container resources.
+
+Thanks to the above:
+* PodSpec.Container.ResourceRequirements becomes purely a declaration,
+  denoting **desired** state of the Pod,
+* PodStatus.ContainerStatus.ResourceAllocated (new object) denotes **actual**
+  state of the Pod resources.
+
+To distinguish between possible states of a Pod resource update,
+a new PodCondition named ResourceResizeRequired is added, with the following states:
```

Comment:
This looks like a state machine, which contradicts our [api conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties).

The general pattern is to have true/false conditions which describe a single aspect of the objects state.  For example PodScheduled true/false.

Reply:
Yes, the condition is being used as a way to communicate state across components (controller, scheduler, kubelet). The conventions doc is slightly self-contradicting. While it calls out the state-machine, it also says "_Conditions should be added to explicitly convey properties that users and components care about rather than requiring those properties to be inferred from other observations_" . That is what we are doing here. 

Let me go over this more closely, but the real question to ask is whether having this PodCondition is meaningful while being simple.

Reply:
I'm not seeing the contradiction.  The quoted text means, for example, that users should not have to separately query the kubelet to ask if a pod is running or not.  That would be "requiring [that property] to be inferred from other observations".  Instead, we place this information in the pod's "Ready" condition.  It is possible to limit conditions to two states (true/false), rather than many states with relations between them (a state machine) and convey properties users and components care about.

As an example (not necessarily what we should do) would be to have conditions:
ResizeScheduled: True if the assigned node meets the scheduling requirements given a pending resize.  False otherwise.
ResizePendingPreemption: True if the resize is waiting for another pod to be removed from the node.  False otherwise.
ResizeFailed: True if any step in the resize has failed for any reason.
ResizeRejected: True if the resize has been permanently rejected.

I see your point about "being simple", but that isn't a good reason to violate API conventions.  I would certainly question the necessity of a few of those conditions.

Reply:
Yes this occurred to me :) .. felt it is overly complicating things, code performance takes a hit..

Reply:
The other option is to have a separate field PodStatus.ResizeState but this may have the same problems as the 'Phase' field the api guide refers to. I'm not privy to the problems that were encountered so I don't know how bad it might be. What do you think?

Reply:
The Phase field is quite a pain, since it requires knowing the current Phase in order to set the new one.  For example, the only way to tell if a pod with no running containers is Pending or Succeeded is to look at the previous Phase.  If it was Pending, then it is Pending.  If it was Running, it must be Succeeded or Failed, etc.  This means whenever the kubelet restarts, it might report that a Succeeded pod is Pending, since it doesn't know the previous Phase immediately after restarting.  So we have to hack around it to reject those status updates, and can get stuck in strange states.

In general, it is a best practice to make the status observable purely based on currently observable properties, rather than based on historical properties.  I.E. PodRunning condition is preferable to the Pending->Running->Succeeded state machine since the state machine's current status is influenced by previous statuses.  The PodRunning condition can always be reconstructed through observation of the actual state of the node.

`ResizeState` and the current condition (with many Reasons) are actually equivalent.  It isn't quite as bad as Phase, since most state transitions are valid, and are recoverable by the system.  But it would still be preferable to represent this with many explicitly defined observations about the state of the world.  `ResizeScheduled` above, for example, is an observation by the scheduler that the desired resources of that pod "fit" on the node.  I'm less sure how one would observe `ResizeFailed` = false, since it is impossible to observe something _not_ happening.  I hope that illustration helps...

Reply:
It does, thanks! 

Let me think more deeply about what its implementation might look like if we broke it down as you laid out above..

Reply:
@dashpole API conventions doc does encourage the use of Reason field. So, how about breaking this down into two conditions as below:

**ResizeSuccess** PodCondition:
  * Status == true    Last resize was successful
  * Status == false   Last resize failed
    * Reason == Rejected    (e.g. resource quota exceeded)
    * Reason == Failed         (e.g. node does not have room)

**ResizePending** PodCondition:
  * Status == false   ResizeSuccess PodCondition shows status of last resize
  * Status == true    A resize is in progress
    * Reason == Awaiting     (e.g. low priority pods being booted)
    * Reason == (empty)      (e.g. Scheduler approved, Kubelet is processing resize)


Scheduler, upon seeing RR != RA and new RR fits node, sets the ResizePending condition to _true/empty_ to tell Kubelet to apply the resource update. Kubelet sets it to false after it is done handling the update, and sets ResizeSuccess condition to either _true_ or _false/Failed_.

These two seems to explicitly communicate pending and done resize statuses to components that are interested, without overloading the condition into a state machine.

Reply:
That is definitely much better, although I have a few more suggestions.  ;)

1. `Reason` shouldn't ever be empty.  There should always be an answer to "Why?" for a state.
2. The T/F of the Condition itself should add information.  We already know a Resize is pending if desired resources and actual resources differ.
3. I would encourage you to be as specific as possible with the reason.  "Awaiting", "Rejected", and "Failed" are all _extremely_ generic answers to "Why?".  Prefer things like "Preempting", "QuotaExceeded", "ContainerResizeFailed" to the above.

To help clarify what i'm after, here is an example of how it _could_ be done.  But you definitely don't need to do this, or anything like this...
**ResourceResizeScheduled** PodCondition:

* `Status == true` means the desired resource requests are deemed valid by the scheduler.
  * `Reason == ValidDesiredResources`
* `Status == false` means the desired resource requests have not been deemed valid by the scheduler.
  * `Reason == NotEnoughRoomOnNode`
  * `Reason == PreemptingPods`
  * `Reason == <new future reason why it can't be scheduled>`

The kubelet only acts on pod resource requests when `ResourceRequestsScheduled == True`.  The scheduler acts on all pods where RR != RA, and ResourceResizeScheduled `Status != True`.   The scheduler continuously tries (with backoff) to "schedule" the new resource requests, just like it would with a normal pod.  If successful, it sets `Status == true`, `Reason == ValidDesiredResources`.  If not, it sets `Status == false`, `Reason == <whatever the reason is>`.

The kubelet acts on the pod once `Status == true`, and re-does admission on the new resource requests.  If that fails, the kubelet rejects the pod, by setting its `Phase == Failed`, and deleting the pod.  This should be exceedingly rare, as the scheduler makes sure it fits beforehand.  The kubelet begins the process of updating the containers, and omits events if any of those steps fail temporarily.  Once it is successful, the kubelet updates the actual resources to be the same as the desired resources, and sets the **ResourceResizeScheduled** PodCondition's `Status == False`, `Reason == NoResizeRequested`.

The controller can react to any reason it chooses to in any way it wishes.  If it decides that it should delete the pod if it runs into `Status == false`, `Reason ==NotEnoughRoomOnNode`, then it is free to do that.

I am sure I missed some design point, so take it with a few grains of salt...  But most importantly, the Condition's T/F value adds important information, and the reasons are just there to explain "Why?".

Reply:
@dashpole Thanks for the review. I'll go over the above breakdown into two distinct conditions with non-empty reasons once more to make sure I didn't miss something, and update the KEP. 

Reasons can themselves be descriptive, or the Message field can carry the full explanation (which is what I initially figured but didn't mention above). I'd perhaps reword the Reasons as "FailedContainerResize", "RejectedQuotaExceeded"... and that way components which need to know the base state (Failed, Rejected, ..) can grep for it to determine their next action. This would also work if we added new reason values in the future.

About Kubelet rejecting pod for failed update, I think that should be policy choice - please see [Retry Policy](https://github.com/kubernetes/enhancements/pull/686/files#diff-57e345a188df5653422eb0ed20bf9442R159) section. Failure to update resource at kubelet is different from failure to admit a new pod. I the resize case, the user may be fine with sub-optimal pod performance (say not enough CPU) as opposed to restart/reschedule which may be costly for a job that has already put in some work. IIRC, this was one of the requirements VPA team requested for in our first design document, but I can imagine Jobs wanting similar choice.

And yes, kubelet will update RA after successful resizing of containers. This is necessary to account for the case with lowering memory that @derekwaynecarr and I discussed in yesetrday's sig-node meeting. I'll update the KEP to reflect this decision.

Reply:
The new conditions look reasonable to me.  Thanks.